### PR TITLE
add precompile-evm docs

### PR DESCRIPTION
**Pull request by ceyonur**
# Docs PR Template

## Why this should be merged

This adds PrecompilEVM repo guide to create precompiles without forking whole Subnet-EVM

## How this works

Refactors Hello World Precompile Tutorial with references to new PrecompilEVM template repo.

## How these changes were tested 

Changes are in https://github.com/ava-labs/precompilevm/tree/hello-world-example

### CODE REVIEW

**Comment by martineckardt** [comment]
In Step 0: I think it makes it overly complicated to have the user clone the empty main branch (there is only a README.md and License) and then have them switch to the helloworld branch. I would suggest to include everything in the main branch except of the helloworld files

Furthermore the tutorial has an error in the branch name. It says the user should checkout 

git checkout -b hello-world-stateful-precompile

but the branch is actually called:

git checkout -b hello-world-example

**Comment by ceyonur** [comment]
> In Step 0: I think it makes it overly complicated to have the user clone the empty main branch (there is only a README.md and License) and then have them switch to the helloworld branch. I would suggest to include everything in the main branch except of the helloworld files

Currently we haven't merged anything to the main branch as we're still reviewing this [PR](https://github.com/ava-labs/precompile-evm/pull/1). But once it merged, everything will be read in main branch (expect hello world).

> Furthermore the tutorial has an error in the branch name. It says the user should checkout
> 
> git checkout -b hello-world-stateful-precompile
> 
> but the branch is actually called:
> 
> git checkout -b hello-world-example

If they clone our repo directly (without forking it) they cannot checkout to a new branch with flag `-b` since there is already a branch named `hello-world-example`. The tutorial also assumes they're checking out to a new branch identical with master (no hello world example) and start adding files.

**Comment by martineckardt** [comment]
Step 13: Building

This only states the build process for subnet-evm, but not for precompile-evm:

Once we've built AvalancheGo, we can navigate back to the Subnet-EVM repo and build the Subnet-EVM binary:

cd $GOPATH/src/github.com/ava-labs/subnet-evm
./scripts/build.sh

It should also give the command:

cd $GOPATH/src/github.com/ava-labs/precompile-evm
./scripts/build.sh

**Comment by martineckardt** [comment]
1.) subnet-evm vs precompile-evm
I think it might be better to separate or restructure. When I went through it, I would follow the tutorial, a step would fail and then I am reading in the next paragraph "This won't work with Precompile-EVM as ....". Depending on how different the processes are, maybe it will be better to separate or at least restructure (maybe tabs for diverging instructions)

2.) Reduce Complexity
I think we can reduce complexity. One thing I would take out is everything about AllowList and move this to an article called "Restricting Access to you Precompile". It is a powerful feature and deserves more room for explaining. Furthermore, I think if someone is creating their first precompile for playing around, this is not necessary and rather something to go to after getting familiar with the concept.

3.) No guide on storing data in the EVM state
When building the academy course I found that the tutorial is missing information on how to store data in the EVM state. I was able to see how it was done for strings in the hello world example, but had to come up with ways to encode and decode to a 32 byte array for other data types such as integers. I believe devs should not have to reinvent the wheel here and we should give them best practices for all data types and common encodings

Most of this I am writing for the academy, so we can easily transfer the contents to the docs. Lmk if you want to see the current progress of the course and I can give you access after you signed up here:
[academy.avax.com](http://academy.avax.com/)

**Comment by ceyonur** [comment]
> 3.) No guide on storing data in the EVM state When building the academy course I found that the tutorial is missing information on how to store data in the EVM state. I was able to see how it was done for strings in the hello world example, but had to come up with ways to encode and decode to a 32 byte array for other data types such as integers. I believe devs should not have to reinvent the wheel here and we should give them best practices for all data types and common encodings
> 

Great feedback! We should have a separate PR for that as it could be a big one. Could you create an issue for that?

I tried to fix the rest of your feedbacks in this PR @martineckardt

File: docs/subnets/hello-world-precompile-tutorial.md [hunk 1]
```diff
@@ -141,14 +166,35 @@ This is a brief overview of what this tutorial will cover.
...
+The difference between using Subnet-EVM and Precompile-EVM is that with Subnet-EVM you can change EVM
+internals to interact with your precompiles.
+Such as changing fee structure, adding new opcodes, changing how to build a block, etc.
+With Precompile-EVM you can only add new stateful precompiles that can interact with the StateDB.
+Precompiles built with Precompile-EVM are still very powerful because it can directly access to the
+state and modify it.
+
+There is a template repo for how to build a precompile with this way called
+[Precompile-EVM](https://github.com/ava-labs/precompile-evm). Both of Subnet-EVM and Precompile-EVM shares
+similar directory structure and common codes.
```
Comment (meaghanfitzgerald) [inline docs/subnets/hello-world-precompile-tutorial.md:192]:
```suggestion
[Precompile-EVM](https://github.com/ava-labs/precompile-evm). Both Subnet-EVM and Precompile-EVM share
similar directory structures and common codes.
```

File: docs/subnets/hello-world-precompile-tutorial.md [hunk 2]
```diff
@@ -141,14 +166,35 @@ This is a brief overview of what this tutorial will cover.
...
+Most of the tutorial still works same way
+for Precompile-EVM repo. Some steps are separated for Subnet-EVM and Precompile-EVM with tabs.
+We followed same Hello World tutorial in Precompile-EVM repo.
+You can access the Precompile-EVM PR that adds Hello World precompile [here](https://github.com/ava-labs/precompile-evm/pull/2)
```
Comment (meaghanfitzgerald) [inline docs/subnets/hello-world-precompile-tutorial.md:196]:
```suggestion
You can reference the Precompile-EVM PR that adds Hello World precompile [here](https://github.com/ava-labs/precompile-evm/pull/2)
```

File: docs/subnets/hello-world-precompile-tutorial.md [hunk 3]
```diff
@@ -141,14 +166,35 @@ This is a brief overview of what this tutorial will cover.
 Stateful precompiles are [alpha software](https://en.wikipedia.org/wiki/Software_release_life_cycle#Alpha).
 Build at your own risk.
```
Comment (meaghanfitzgerald) [inline docs/subnets/hello-world-precompile-tutorial.md:167]:
```suggestion
:::caution
Stateful precompiles are [alpha software](https://en.wikipedia.org/wiki/Software_release_life_cycle#Alpha).
Build at your own risk.
:::
```

File: docs/subnets/customize-a-subnet.md [hunk 4]
```diff
@@ -825,8 +828,8 @@ Configurations](#initial-precompile-configurations).
 
 ## Contract Examples
 
-Subnet-EVM contains example contracts for precompiles under `/contract-examples`. It's a hardhat
-project with tests, tasks. For more information see [contract examples README](https://github.com/ava-labs/subnet-evm/tree/master/contract-examples#subnet-evm-contracts).
+Subnet-EVM contains example contracts for precompiles under `/contracts`. It's a hardhat
+project with tests, tasks. For more information see [contract examples README](https://github.com/ava-labs/subnet-evm/tree/master/contracts#subnet-evm-contracts).
```
Comment (meaghanfitzgerald) [inline docs/subnets/customize-a-subnet.md:832]:
```suggestion
Subnet-EVM contains example contracts for precompiles under `/contracts`. It's a hardhat
project with tests and tasks. For more information see 
[contract examples README](https://github.com/ava-labs/subnet-evm/tree/master/contracts#subnet-evm-contracts).
```

**Review by meaghanfitzgerald — COMMENTED** [review]
nits and general comments, afaict LGTM.

**Comment by meaghanfitzgerald** [comment]
I merged a [change](https://github.com/ava-labs/avalanche-docs/commit/b9faeac36b609405a6e5aa0e05d3c2bfdfaf65ab) i wanted made to master on the instructions for setting the GOPATH depending on os . Since @ceyonur already acknowledged @martineckardt's comments above about complexity issues, we should do our best to merge this tutorial as it is now and plan to reduce complexities and compartmentalize steps in a larger PR in the future (like (https://github.com/ava-labs/avalanche-docs/pull/1449))

The future PR structure should follow the structure referenced in the docs revamp Mural:
image

Each red comment has linked an existing file in (https://github.com/ava-labs/avalanche-docs/pull/1449) to the structure.

Great work everyone 💯

**Review by meaghanfitzgerald — APPROVED** [review]

**Pull request merged**